packages/ifpack2/src/Ifpack2_Details_AdditiveSchwarzFilter_def.hpp

```diff
@@ -0,0 +1,769 @@
+          for(size_type j = rowBegin; j < rowEnd; j++)
+          {
+            local_ordinal_type entry = A_halo.graph.entries(j);
+            if(entry < totalLocalRows && permView(entry) != -1)
+              numEntries++;
+          }
+        }
+        localRowptrs(i) = numEntries;
+        if(i == numLocalRows - 1)
+          localRowptrs(i + 1) = 0;
+      });
+  //Prefix sum to finish computing the rowptrs
+  size_type numLocalEntries;
+  Kokkos::parallel_scan(policy_type(0, numLocalRows + 1),
+      KOKKOS_LAMBDA(local_ordinal_type i, size_type& lnumEntries, bool finalPass)
+      {
+        size_type numEnt = localRowptrs(i);
+        if(finalPass)
+          localRowptrs(i) = lnumEntries;
+        lnumEntries += numEnt;
+      }, numLocalEntries);
+  //Allocate and fill local entries and values
+  entries_type localEntries(Kokkos::view_alloc(Kokkos::WithoutInitializing, "Filtered entries"), numLocalEntries);
+  values_type localValues(Kokkos::view_alloc(Kokkos::WithoutInitializing, "Filtered values"), numLocalEntries);
+  //Create the local matrix with the uninitialized entries/values, then fill it.
+  local_matrix_type localMatrix("AdditiveSchwarzFilter", numLocalRows, numLocalRows, numLocalEntries, localValues, localRowptrs, localEntries);
+  fillLocalMatrix(localMatrix);
+  //Create a serial comm and the map for the final filtered CrsMatrix (each process uses its own local map)
+#ifdef HAVE_IFPACK2_DEBUG
+  TEUCHOS_TEST_FOR_EXCEPTION(
+    ! mapPairsAreFitted (*A_unfiltered_), std::invalid_argument, "Ifpack2::LocalFilter: "
+    "A's Map pairs are not fitted to each other on Process "
+    << A_->getRowMap ()->getComm ()->getRank () << " of the input matrix's "
+    "communicator.  "
+    "This means that LocalFilter does not currently know how to work with A.  "
+    "This will change soon.  Please see discussion of Bug 5992.");
+#endif // HAVE_IFPACK2_DEBUG
+
+  // Build the local communicator (containing this process only).
+  RCP<const Teuchos::Comm<int> > localComm;
+#ifdef HAVE_MPI
+  localComm = rcp (new Teuchos::MpiComm<int> (MPI_COMM_SELF));
+#else
+  localComm = rcp (new Teuchos::SerialComm<int> ());
+#endif // HAVE_MPI
+  //All 4 maps are the same for a local square operator.
+  localMap_ = rcp (new map_type (numLocalRows, 0, localComm, Tpetra::GloballyDistributed));
+  //Create the inner filtered matrix.
+  auto crsParams = rcp(new Teuchos::ParameterList);
+  crsParams->template set<bool>("sorted", true);
+  //NOTE: this is the fastest way to construct A_ - it's created as fillComplete,
+  //and no communication needs to be done since localMap_ uses a local comm.
+  //It does need to copy the whole local matrix to host when DualViews are constructed
+  //(only an issue with non-UVM GPU backends) but this is just unavoidable when creating a Tpetra::CrsMatrix.
+  //It also needs to compute local constants (maxNumRowEntries) but this should be a
+  //cheap operation relative to what this constructor already did.
```

Comment:
My understanding:   This copy-to-host only happens once, at construction.  The matrix graph is fixed, and if matrix values are later updated, no copy to host occurs.

Reply:
@jhux2 Unfortunately it also copies when I do setAllValues, since that re-constructs the WrappedDualViews in the matrix and graph. I don't think there's any other legal way to replace all the values without doing resumeFill/fillComplete, which I wanted to avoid.

Reply:
Oh, that's too bad.  Perhaps fodder for discussion at a future Tpetra meeting.

Reply:
In future, if Karen's suggestion #9655 gets implemented, that would make this really easy because the WrappedDualView will handle it automatically. I already get a ReadWrite view of the matrix values, so I can just change them and then be done. No copying to host or fillComplete needed.